File: leetCode/python/p30_SubstringWithConcatenationOfAllWords/lc30.py
#!/usr/bin/python

class Solution:
	"""
	You are given a string s and an array of strings words. All the strings of words are of the same length.
	
	A concatenated substring in s is a substring that contains all the strings of any permutation of words concatenated.
	
	For example, if words = ["ab","cd","ef"], then "abcdef", "abefcd", "cdabef", "cdefab", "efabcd", and "efcdab" are all concatenated strings. 
	"acdbef" is not a concatenated substring because it is not the concatenation of any permutation of words.
	Return the starting indices of all the concatenated substrings in s. You can return the answer in any order.
	"""

	def clear_words_from_dict(self, words_cnt: dict):
		
		for word in words_cnt:
			words_cnt[word] = 0

	def check_list_completion(self, words_cnt: dict):
		
		dict_size = len(words_cnt)
		num_of_word_inst = 0

		for word in words_cnt:
			
			if words_cnt[word] == 1:
				num_of_word_inst += 1
			
			if words_cnt[word] > 1:
				return False

		if num_of_word_inst == dict_size:
			return True
		else:
			return False
	
	def findSubstring(self, s: str, words: list) -> list:
		
		result = list()

		# Check if string exists
		if s is None or len(s) == 0 or words is None or len(words) == 0:
			return result

		words_cnt = dict()
	
		for word in words:
			words_cnt[word] = 0

		word_length = len(words[0])

		word_array_length = len(words) * word_length

		for i in range(len(s) - word_array_length + 1):
			
			self.clear_words_from_dict(words_cnt)
			word = s[i:i+3]		
	
			if word in words_cnt:
	
				words_cnt[word] += 1
		
				j = i + 3

				while s[j:j+3] in words_cnt:

					if words_cnt[s[j:j+3]] == 0:
							
						words_cnt[s[j:j+3]] += 1	
							
						if self.check_list_completion(words_cnt):
							result.append(i)
							i = i + 2
							break

						else:		
							j += 3	
					
					else:
						break
				

		return result
		
s = Solution()

# Test 1:

s1 = 'barfoothefoobarman'
words1 = ['foo','bar']

print(s.findSubstring(s1, words1))

# Test 2:

s2 = 'wordgoodgoodgoodbestword'
words2 = ['word','good','best','word']

print(s.findSubstring(s2, words2))

# Test 3:

s3 = 'barfoofoobarthefoobarman'
words3 = ['bar', 'foo', 'the']

print(s.findSubstring(s3, words3))

File: leetCode/cpp/p30_SubstringWithConcatenationOfAllWords/lc30.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> findSubstring(string s, vector<string> &words);    
    void printVector(vector<string>& vec);
    void printVector(vector<int>& vec);
};

void Solution::printVector(vector<string>& vec) {
   
    for(string str: vec) 
        cout << str << ' ';

    cout << '\n';
}

void Solution::printVector(vector<int>& vec) {
   
    for(int str: vec) 
        cout << str << ' ';

    cout << '\n';
}

vector<int> Solution::findSubstring(string s, vector<string> &words) {

    unsigned int k = words.size();
    unsigned int l = words[0].length();
    vector<int> ans;
    unordered_map<string, int> count;

    if(s.empty() || words.empty())
        return {};
    
    for(const string& word : words)
        count[word]++;

    for(int i=0; i < s.length() - k * l + 1; i++) {
        unordered_map<string, int> seen;
        int j;

        for(j=0; j < k; j++) {
            const string& word = s.substr(i + j * l, l);
            seen[word]++;
 
            if(seen[word] > count[word])
                break;            
        }

        if(j == k)
            ans.push_back(i);
    }
    
    return ans;
}

int main(void) {

    Solution s;
        
    /* Test 1: */
    cout << "Test 1:\n";

    string str1 = "barfoothefoobarman";
    vector<string> words1 = {"foo", "bar"};

    cout << "String 1: " << str1 << '\n';
    cout << "Words 1: ";
    s.printVector(words1);

    vector<int> sol1 = s.findSubstring(str1, words1);
  
    cout << "Solution 1: ";
    s.printVector(sol1);
    cout << '\n';

    /* Test 2: */
    cout << "Test 2:\n";

    string str2 = "wordgoodgoodgoodbestword";
    vector<string> words2 = {"word", "good", "best", "word"};

    cout << "String 2: " << str2 << '\n';
    cout << "Words 2: ";
    s.printVector(words2);

    vector<int> sol2 = s.findSubstring(str2, words2);
  
    cout << "Solution 2: ";
    s.printVector(sol2);
    cout << '\n';

    /* Test 3: */
    cout << "Test 3:\n";

    string str3 = "barfoofoobarthefoobarman";
    vector<string> words3 = {"bar", "foo", "the"};

    cout << "String 3: " << str3 << '\n';
    cout << "Words 3: ";
    s.printVector(words3);

    vector<int> sol3 = s.findSubstring(str3, words3);
  
    cout << "Solution 3: ";
    s.printVector(sol3);
    cout << '\n';
    
    return EXIT_SUCCESS; 
}
